### PR TITLE
Add support for CAN FMI on STM32

### LINKED ISSUE #225: Can filter match information is lost in STM32 can driver
The Filter match index (`FMI`) from "**CAN receive FIFO mailbox data length control and time stamp register**"(`CAN_RDTxR`) is not copied over to `Can::Message`

### PR DESCRIPTION
**Pull request by asmfreak**
Fixes #225 .
This is only done for STM32. I'm not sure how to make filters work on other platforms.

### CODE REVIEW

**Review by salkinium — DISMISSED** [review]
Thanks!

**Comment by salkinium** [comment]
The FMI is pretty STM32 specific anyways, so that's fine.

**Comment by dergraaf** [comment]
Does this make sense in the message class? The filter is meaningless when sending messages. I think it should be something which you get as additional information when receiving the message.

**Comment by salkinium** [comment]
Something backwards compatible like this?
```cpp
static bool
getMessage(can::Message& message, uint8_t *filter_id=nullptr);

can::Message m;
uint8_t id;
if (Can1::getMessage(m, &id)) {
    // do something
}
```

File: src/modm/architecture/interface/can_message.hpp [hunk 1]
```diff
@@ -94,6 +100,7 @@ struct Message
 		bool extended : 1;
 	} flags;
 	uint8_t length;
+    uint8_t filter_id;
```
Comment (dergraaf) [inline src/modm/architecture/interface/can_message.hpp]: I don't think this makes sense here. This is not a property of a message, but rather an artifact during reception of a message.

File: src/modm/platform/can/stm32/can.cpp.in [hunk 2]
```diff
@@ -201,6 +201,7 @@ readMailbox(modm::can::Message& message, uint32_t mailboxId)
 	message.setRemoteTransmitRequest(rir & CAN_RI0R_RTR);
 
 	message.length = mailbox->RDTR & CAN_TDT1R_DLC;
+    message.filter_id = (mailbox->RDTR & CAN_RDT1R_FMI) >> CAN_RDT1R_FMI_Pos;
```
Comment (dergraaf) [inline src/modm/platform/can/stm32/can.cpp.in]: The `getMessage` function might be a better place to return this value. E.g. as an optional reference parameter.

**Review by dergraaf — DISMISSED** [review]

**Comment by salkinium** [comment]
The atomic queue inside the CAN implementation then needs to have a struct type. Something like:
```cpp
struct RxMessage {
    modm::can::Message message;
    uint8_t filter_id;
};
static modm::atomic::Queue<RxMessage, {{ options["buffer.rx"] }}> rxQueue;
```

**Comment by dergraaf** [comment]
@salkinium 
You're too fast in responding :smile:. Yes something like this would be my suggestion as well.

**Comment by salkinium** [comment]
> You're too fast in responding 😄.

I get push notifications of all events via [Canopy](https://mxcl.dev/canopy/). I even see when you edit your comments. Mildy creepy…

**Comment by asmfreak** [comment]
@salkinium 
Should I change this signature in all can drivers?
```
static bool
getMessage(can::Message& message, uint8_t *filter_id=nullptr);
```

**Comment by salkinium** [comment]
No, just the signature of the STM32 implementation. With the nullptr default argument it is backwards compatible. The Filter ID is STM32 specific if I understand correctly.

**Comment by asmfreak** [comment]
@salkinium 
Yep, just as I thought.

**Comment by asmfreak** [comment]
I've also modified example to use this change.

**Review by salkinium — APPROVED** [review]
Much better now!

**Comment by salkinium** [comment]
I've squashed and rebased your commits on develop.

**Pull request merged**